### ISSUE TITLE
SplitText is stupidly slow to split a file in large chunks

### ISSUE THREAD

**Issue by lucidv01d** [nextflow-io/nextflow#102]
I'm trying to run a nextflow script on our SGE cluster but it keeps hanging at the [warm up] stage:
```
N E X T F L O W  ~  version 0.17.0
Launching sim_vep.nf
[warm up] executor > sge
```
Since other people are not having issues with SGE executors, I think the issue must be with our cluster. How can I go about debugging this? 

My nextflow log looks like this:
```
Jan-15 16:54:39.431 [main] DEBUG nextflow.cli.Launcher - $> nextflow/nextflow run sim_vep.nf
Jan-15 16:54:39.676 [main] INFO  nextflow.cli.CmdRun - N E X T F L O W  ~  version 0.17.0
Jan-15 16:54:39.677 [main] INFO  nextflow.cli.CmdRun - Launching sim_vep.nf
Jan-15 16:54:40.457 [main] DEBUG nextflow.Session - Session uuid: c70dcad7-5df3-485d-bc1f-1fed331a0a1a
Jan-15 16:54:40.463 [main] DEBUG nextflow.Session - Executor pool size: 17
Jan-15 16:54:40.481 [main] DEBUG nextflow.cli.CmdRun -
  Version: 0.17.0 build 3441
  Modified: 16-01-2016 00:48 UTC (15-01-2016 16:48 PDT)
  System: Linux 2.6.32-504.12.2.el6.664g0000.x86_64
  Runtime: Groovy 2.4.5 on Java HotSpot(TM) 64-Bit Server VM 1.7.0_72-b14
  Encoding: UTF-8 (UTF-8)
  Process: 468392@cclc01.som.ucsf.edu [10.54.0.1]

Jan-15 16:54:40.504 [main] DEBUG nextflow.Session - Session start invoked
Jan-15 16:54:40.521 [main] DEBUG nextflow.processor.TaskDispatcher - Dispatcher > start
Jan-15 16:54:40.523 [main] DEBUG nextflow.script.ScriptRunner - > Script parsing
Jan-15 16:54:41.071 [main] DEBUG nextflow.script.ScriptRunner - > Launching execution
Jan-15 16:54:41.713 [main] DEBUG nextflow.processor.ProcessFactory - << taskConfig executor: sge
Jan-15 16:54:41.714 [main] DEBUG nextflow.processor.ProcessFactory - >> processorType: 'sge'
Jan-15 16:54:41.756 [main] DEBUG nextflow.executor.Executor - Initializing executor: sge
Jan-15 16:54:41.762 [main] INFO  nextflow.executor.Executor - [warm up] executor > sge
Jan-15 16:54:41.804 [main] DEBUG n.processor.TaskPollingMonitor - Creating task monitor for executor 'sge' > capacity: 100; pollInterval: 1s; dumpInterval: 5m
Jan-15 16:54:41.827 [main] DEBUG nextflow.processor.TaskDispatcher - Starting monitor: TaskPollingMonitor
Jan-15 16:54:41.829 [main] DEBUG n.processor.TaskPollingMonitor - >>> barrier register (monitor: sge)
Jan-15 16:54:41.848 [main] DEBUG nextflow.executor.Executor - Invoke register for executor: sge
Jan-15 16:54:41.851 [main] DEBUG n.executor.AbstractGridExecutor - Creating executor 'sge' > queue-stat-interval: 1m
Jan-15 16:54:41.935 [main] DEBUG nextflow.Session - >>> barrier register (process: vep)
Jan-15 16:54:41.947 [main] DEBUG nextflow.script.BaseParam - output > channel unknown: annotated -- creating a new instance
Jan-15 16:54:41.990 [main] DEBUG n.processor.ParallelTaskProcessor - Creating operator > vep -- maxForks: 17
Jan-15 16:54:42.078 [main] DEBUG nextflow.script.ScriptRunner - > Await termination
Jan-15 16:54:42.079 [main] DEBUG nextflow.Session - Session await
Jan-15 16:59:43.039 [Thread-2] DEBUG n.processor.TaskPollingMonitor - !! executor sge > tasks to be completed: 0 -- first: null
Jan-15 17:04:43.192 [Thread-2] DEBUG n.processor.TaskPollingMonitor - !! executor sge > tasks to be completed: 0 -- first: null
```

**Comment by pditommaso** [comment]
Did you tried to run it locally (I mean without SGE)? I guess that something wrong with your script. Could you shared it?

**Comment by lucidv01d** [comment]
The script runs fine locally, though there are some path changes. On the cluster, the script actually does complete, but only after 1.5 hours in the warm up stage. Do the 'Submitted process ...' messages show up when the job starts running or when first submitted to the queue? I don't see any jobs submitted under my username with **qstat** during the wait time, so I'm not sure the behavior. 

Thanks for the help by the way, absolutely love nextflow.

**Comment by pditommaso** [comment]
The message `Submitted process` shows up when the job is submitted with `qsub` command. That's really weird. Could you launch nextflow in debug mode by using the following command:

    nextflow -trace nextflow run <your script> ..

**Comment by lucidv01d** [comment]
The debug trace seemed to show the script caught in a loop outputting: 
```
TRACE: Scheduler queue size: 0
```
so I tried submitting to another queue that we have called *batch* (we normally don't have to specify this queue). That didn't work either, so I down-sampled my input file and this got it past the warm up stage. Should the splitText operation take ~1.5 hour for around 12 splits of a 6 million line file?

Even with the down-sampled file, the script still bails with the following error:
``` 
qsub: submit error (Bad Job Array Request)
``` 

Not sure what to do; googling doesn't reveal much. 

Here is my script. I've tried as is and with the memory and queue directives commented out:
```
#!/usr/bin/env nextflow
params.input	  = "$HOME/simulated_variants.vcf"
params.output     = "$HOME/simulated_variants_annotated.vcf"
vep_cmd           = "/software/ensembl-tools-release-83/scripts/variant_effect_predictor/variant_effect_predictor.pl"
vep_flags         = "--everything --dont_skip --offline --cache --assembly GRCh37 --force_overwrite"
params.chunkSize  = 500000

Channel
    .fromPath(params.input)
    .splitText(by: params.chunkSize)
    .set { vcfs }

process vep {
    executor 'sge'
    queue 'batch'
    memory '10 GB'

    input:
    file 'unannotated' from vcfs

    output:
    file 'annotated'

    """
    $vep_cmd $vep_flags -i unannotated -o annotated
    """
}

annotated
    .collectFile(name: params.output)
```

**Comment by pditommaso** [comment]
umh, I need to investigate that. However an easy workaround is to add `file: true` option to the `splitText` operator. For example: 

    Channel
        .fromPath(params.input)
        .splitText(by: params.chunkSize, file: true)
        .set { vcfs }

The difference is that it creates the chunks to files instead of of retain them in memory. 


Reading the error: 

    qsub: submit error (Bad Job Array Request)

This is definitely a different problem. Could you please open a separate issue including the `.nextflow.log` file obtained running with the `-trace nextflow` option?

**Comment by pditommaso** [comment]
OK, I've found the problem and pushed a patch. It will be included in the next release.

**Comment by lucidv01d** [comment]
Awesome, the ```file: true``` option works great. Attached is my ```.nextflow.log``` file 
[nextflow.zip](https://github.com/nextflow-io/nextflow/files/96872/nextflow.zip)

**Comment by pditommaso** [comment]
Regarding the `qsub` error are you sure that you are using a SGE cluster? Googling for that error message it seems that is produced by PBS/Torque. I would try to use `executor 'pbs'`.

**Comment by lucidv01d** [comment]
I was misinformed, the cluster is indeed PBS/Torque. Working beautifully now, thanks for all the help!

**Comment by pditommaso** [comment]
Enjoy!

**Comment by pditommaso** [comment]
Fixed with version `0.17.1`
